bootstrap/src/stream.ml

```diff
@@ -0,0 +1,576 @@
+open Rudiments
+
+type 'a elm =
+  | Nil
+  | Cons of 'a * 'a t
+and 'a t = 'a elm lazy_t
+
+let empty = lazy Nil
+
+let init n ~f =
+  let rec fn i n f = lazy begin
+    match i < n with
+    | false -> Nil
+    | true -> begin
+      let elm = f i in
+      let t = fn (Usize.succ i) n f in
+      Cons(elm, t)
+    end
+  end in
+  fn 0 n f
+
+let length t =
+  let rec fn t i = begin
+    match t with
+    | (lazy Nil) -> i
+    | (lazy (Cons(_, t'))) -> fn t' (Usize.succ i)
+  end in
+  fn t 0
+
+let is_empty = function
+  | lazy Nil -> true
+  | lazy (Cons(_, _)) -> false
+
+let hd = function
+  | lazy Nil -> halt "Empty stream has no head"
+  | lazy (Cons(elm, _)) -> elm
+
+let tl = function
+  | lazy Nil -> halt "Empty stream has no tail"
+  | lazy (Cons(_, t)) -> t
+
+let push' elm s =
```

Comment:
What do you think about renaming this to `cons`?  `push'` initially made me think it was an internal function just for `push`.

Hmm, on second thought after reading through later code, I like the consistency of the `foo'` functions being non-lazy counterparts to lazy `foo` functions.  It might be worth a comment above this function to call out the convention though.